### ISSUE TITLE
Specify and implement type casting rules

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#157]
Commit 5769347ebc7669c12476029254fc34b491c6aaca added type casting capabilities to the Ruby compiler, but without any rules stating when types can be casted from A to B, or when they can not. This means you can currently freely cast a string to an Array of integers, which isn't exactly what I want.

Some of the rules I can think of currently:

* Casting from/to Dynamic is always allowed
* Casting from an optional to a non-optional is not allowed (other than casting to Dynamic). This to prevent you from casting `?Integer` to `Integer` when you really have a Nil value.
* Casting to Array, Block, or HashMap is only allowed if the source type is the same (e.g. `some_hash as HashMap!(A, B)` is fine but `some_string as HashMap!(A, B)` is not).
* Casting from a generic type (e.g. `Array!(A)`) to a non-generic type (e.g. just `Array`) is not allowed.

Basically the casting rules should enforce that we don't lose information (except when casting to Dynamic), or that we don't add information based on nothing (e.g. casting a String to `Array!(T)` which makes no sense).
